Applications/Spire/Source/Ui/OverlayPanel.cpp

```diff
@@ -0,0 +1,219 @@
+#include "Spire/Ui/OverlayPanel.hpp"
+#include <QApplication>
+#include <QCloseEvent>
+#include <QEvent>
+#include <QGraphicsDropShadowEffect>
+#include <QGuiApplication>
+#include <QHBoxLayout>
+#include <QPropertyAnimation>
+#include <QScreen>
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Ui/Box.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  const auto DROP_SHADOW_COLOR = QColor(0, 0, 0, 64);
+  const auto DROP_SHADOW_OFFSET = QPoint(0, 3);
+  const auto DROP_SHADOW_RADIUS = 5;
+  const auto DROP_SHADOW_SIZE = 3;
+  const auto FADE_SPEED_MS = 100;
+
+  auto DROP_SHADOW_HEIGHT() {
+    static auto height = scale_height(DROP_SHADOW_SIZE);
+    return height;
+  }
+
+  auto DROP_SHADOW_WIDTH() {
+    static auto width = scale_width(DROP_SHADOW_SIZE);
+    return width;
+  }
+
+  auto DROP_SHADOW_MARGINS() {
+    static auto margins = QMargins(DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT(),
+      DROP_SHADOW_WIDTH(), DROP_SHADOW_HEIGHT());
+    return margins;
+  }
+
+  auto DEFAULT_STYLE() {
+    auto style = StyleSheet();
+    style.get(Any()).
+      set(BackgroundColor(QColor::fromRgb(0xFF, 0xFF, 0xFF))).
+      set(border(scale_width(1), QColor::fromRgb(0xA0, 0xA0, 0xA0)));
+    return style;
+  }
+
+  QWidget* find_top_level_window() {
+    auto list = QApplication::topLevelWidgets();
+    for(auto i = 0; i < list.size(); ++i) {
+      auto widget = list.at(i);
+      if(widget->isVisible() && !widget->parentWidget() && widget->isWindow()) {
+        return widget;
+      }
+    }
+    return nullptr;
+  }
+}
+
+OverlayPanel::OverlayPanel(QWidget* body, QWidget* parent)
+    : QWidget(parent, Qt::Tool | Qt::FramelessWindowHint |
+        Qt::NoDropShadowWindowHint),
+      m_body(body),
+      m_top_level_window(nullptr),
+      m_is_closed_on_blur(true),
+      m_is_closed(false),
+      m_positioning(Positioning::PARENT) {
+  setAttribute(Qt::WA_TranslucentBackground);
+  setAttribute(Qt::WA_QuitOnClose);
```

Comment:
Why is this attribute set?

Reply:
It makes Qt quit the application when the last widget with the attribute set has accepted closeEvent(). If the attribute is not set and the application is closed, the GUI will disappear but the application will be alive to be a zombie process.

Reply:
Sounds good.